compiler/rustc_builtin_macros/src/format.rs

```diff
@@ -711,11 +711,11 @@ fn report_missing_placeholders(
                     };
 
                     let pos = sub.position();
-                    let sub = String::from(sub.as_str());
-                    if explained.contains(&sub) {
+                    let sub = sub.as_str();
+                    if explained.contains(sub.as_ref()) {
                         continue;
                     }
-                    explained.insert(sub);
+                    explained.insert(String::from(sub));
```

Comment:
If we're optimizing for readability here, I'd probably write this like:

```rust
if !explained.insert(sub.as_str()) {
    continue;
}
```

Reply:
And possibly also change the `as_str` to call it `to_string` to make it clear that it returns a `String`.